dash-renderer/src/persistence.js

```diff
@@ -0,0 +1,497 @@
+ * - `apply` puts an extracted value back into the prop. Make sure this creates
+ *   a new object rather than mutating `proValue`, and that if there are
+ *   multiple `piece` entries for one `propName`, their `apply` functions
+ *   commute - which should not be an issue if they extract and apply
+ *   non-intersecting parts of the full prop.
+ * You only need to define these for the props that need them.
+ * It's important that `extract` pulls out *only* the relevant pieces of the
+ * prop, because persistence is only maintained if the extracted value of the
+ * prop before applying persistence is the same as it was before the user's
+ * changes.
+ */
+
+import {
+    equals,
+    filter,
+    forEach,
+    keys,
+    lensPath,
+    set,
+    symmetricDifference,
+    type,
+} from 'ramda';
+import {createAction} from 'redux-actions';
+import uniqid from 'uniqid';
+
+import Registry from './registry';
+
+const storePrefix = '_dash_persistence.';
+const UNDEFINED = 'U';
+
+function err(e) {
+    const error = typeof e === 'string' ? new Error(e) : e;
+
+    /* eslint-disable no-console */
+    // Send this to the console too, so it's still available with debug off
+    console.error(e);
+    /* eslint-disable no-console */
+
+    return createAction('ON_ERROR')({
+        myUID: uniqid(),
+        myID: storePrefix,
+        type: 'frontEnd',
+        error,
+    });
+}
+
+/*
+ * Does a key fit this prefix? Must either be an exact match
+ * or, if a separator is provided, a scoped match - exact prefix
+ * followed by the separator (then anything else)
+ */
+function keyPrefixMatch(prefix, separator) {
+    const fullStr = prefix + separator;
+    const fullLen = fullStr.length;
+    return key => key === prefix || key.substr(0, fullLen) === fullStr;
+}
+
+class WebStore {
+    constructor(backEnd) {
+        this._name = backEnd;
+        this._storage = window[backEnd];
+    }
+
+    hasItem(key) {
+        return this._storage.getItem(storePrefix + key) !== null;
+    }
+
+    getItem(key) {
+        // note: _storage.getItem returns null on missing keys
+        // and JSON.parse(null) returns null as well
+        const gotVal = this._storage.getItem(storePrefix + key);
+        return gotVal === UNDEFINED ? void 0 : JSON.parse(gotVal);
+    }
+
+    /*
+     * In addition to the regular key->value to set, setItem takes
+     * dispatch as a parameter, so it can report OOM to devtools
+     */
+    setItem(key, value, dispatch) {
+        const setVal = value === void 0 ? UNDEFINED : JSON.stringify(value);
+        try {
+            this._storage.setItem(storePrefix + key, setVal);
+        } catch (e) {
+            if (dispatch) {
+                dispatch(err(e));
+            } else {
+                throw e;
+            }
+            // TODO: Should we clear storage here? Or fall back to memory?
+            // Probably not, unless we want to handle this at a higher level
+            // so we can keep all 3 items in sync
```

Comment:
I think this is fine for our first implementation of persistence.
1. Make sure there's some space available in storage, clear if not, fallback to memory if still not.
2. Write until there's no space left.

Not sure what the internal error looks like but maybe we want to make it cleaner / user focused, like in the initialization step.

Reply:
> Not sure what the internal error looks like but maybe we want to make it cleaner / user focused, like in the initialization step.

good point - gave it a nicer error message -> https://github.com/plotly/dash/pull/903/commits/3669df63e7b9806fdafa6e22111a7b05d76762c0